ASoC:sof:hda: fix hda preload issue

RanderWang opened the pull request:
This issue could be reproduced by multiplex_pipeline-test.sh with two
pipelines. The hda preload would be in unexpected status sometimes 
and this results ipc timeout. It is a HW restriction and the solution is 
advised by Marcin:

1. Put DMA into coupled mode by clearing PPCTL.PROCEN bit
    for corresponding stream index before the time of writing
    format to SDxFMT register.
2. Write SDxFMT
3. Set PPCTL.PROCEN bit for corresponding stream index to
    enable decoupled mode

I reproduced this issue on Apollolake UP2 a few times and validated the fix
a few times. I can't reproduce it on Icelake because this platform is not
stable with multiple threads test for other reasons, but this patch
didn't make any negative effect with single thread test of playback
or capture.

It fixes https://github.com/thesofproject/linux/issues/670 and https://github.com/thesofproject/sof/issues/1050
Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>
Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

keyonjie commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -363,12 +363,30 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
+
+	/* disable DSP features */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR, SOF_HDA_REG_PP_PPCTL,
+				SOF_HDA_PPCTL_GPROCEN, 0);
```
Is this must according to your test? In Marcin's description, only PROCEN(for corresponding stream index) need to be set.

RanderWang replied:
each stream would call this function, right ?

RanderWang replied:
The issue could be fixed without this code

keyonjie replied:
right, each stream will call this.

plbossart replied:
As @keyonjie mentioned it, it's very odd that we would use the general bit, that might impact already running streams. Please use PROCEN.

RanderWang replied:
yes, PROCEN is better and I will refine and test it.

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -363,12 +363,30 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 
+	/* couple host and link DMA */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR,
+				SOF_HDA_REG_PP_PPCTL,
```
for consistency with the code below move this line above so that you have a pattern BAR, register <newline>, mask, value.

RanderWang replied:
yes, I will refine it

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -363,12 +363,30 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 			  sd_offset + SOF_HDA_ADSP_REG_CL_SD_CBL,
 			  hstream->bufsize);
 
```
The commit message is quite good :-)
I would like the sequence to be inserted in the code as comments so that we have a trace of the hardware work-around even after this patch is squashed. 
Thanks!

RanderWang replied:
yes, I will embed the sequence as comments in our code.

RanderWang commented:
update my PR and tested it on Apollolake UP2.

Clarify the solution by Marcin: clearing PPCTL.PROCEN bit for a stream  would decouple HDA host & link dma for that stream, so I refine the solution to 3 steps.

RanderWang commented:
please hold on for I find a ipc timeout after 780 cycles. I will test it tonight.

RanderWang commented:
Validate it a few times again : without this patch, IPC got timeout less than 200 cycles of start-exit  while with it, UP2 survived in 1000 cycles test.  So this patch is effective.

keyonjie commented:
> Validate it a few times again : without this patch, IPC got timeout less than 200 cycles of start-exit while with it, UP2 survived in 1000 cycles test. So this patch is effective.

So are you meaning that the patch can only improve(from <200 to 1000 iterations passed) but doesn't eliminate the timeout issue? 

Please do more stress test to confirm that, if configuring SOF_HDA_PPCTL_GPROCEN is actually needed according to your test, let's change the comment to mention about that also.

lgirdwood commented:
@RanderWang see https://github.com/thesofproject/sof/issues/1050 and align, this also may need FW changes too.

RanderWang commented:
> > Validate it a few times again : without this patch, IPC got timeout less than 200 cycles of start-exit while with it, UP2 survived in 1000 cycles test. So this patch is effective.
> 
> So are you meaning that the patch can only improve(from <200 to 1000 iterations passed) but doesn't eliminate the timeout issue?
> 
> Please do more stress test to confirm that, if configuring SOF_HDA_PPCTL_GPROCEN is actually needed according to your test, let's change the comment to mention about that also.

The maximum loop count in the test file is 1000,  so it would stop at 1000.

RanderWang commented:
@keyonjie  I modify the script and the UP2 is still alive after 1700 cycles.

plbossart commented:
@lgirdwood the commit message states this PR fixes issues #1050 so I will trust Rander's statement.

plbossart commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -363,12 +363,30 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 			  hstream->bufsize);
 
+	/*
+	 * Fix for HDA preload issue due to HW restriction
```
there are never hw restrictions, only undocumented features :-)

Please change to : recommended hardware programming sequence for HDAudio DMA format.

RanderWang replied:
thank, this is a topic our team has discussed how to deal with this type of issue.
This is another solution.

RanderWang commented:
> @lgirdwood the commit message states this PR fixes issues #1050 so I will trust Rander's statement.

Yes, I reproduced 1050 and fixed it, so I added it to my list.

wenqingfu commented:
> > @lgirdwood the commit message states this PR fixes issues #1050 so I will trust Rander's statement.
> 
> Yes, I reproduced 1050 and fixed it, so I added it to my list.

Looks like the 'fixes' keyword can only close on Issue. Since your patch merged, please follow up with a comment and close thesofproject/sof#1050 accordingly, thanks!
